File: data/typescript.cc/GoodsList.ts
import Packet from '../net/Packet';
import GoodsItem from './GoodsItem';


export default class GoodsList {
	private _goods: GoodsItem[] = [];

	constructor(packet?: Packet) {
		if (packet) {
			this._goods = [];
			let goods_count: number = packet.ReadUshort();
			for (let i: number = 0; i < goods_count; i++) {
				this._goods.push(new GoodsItem(packet));
			}
		}
	}

	public Encode(): Packet {
		const packet = this._encode();
		packet.Encode(3020);
		return packet;
	}

	public GetBuffer(): ByteBuffer {
		return this._encode().GetBuffer();
	}

	private _encode(): Packet {
		let packet: Packet = new Packet();
		let goods_count: number = this._goods.length;
		packet.WriteUshort(goods_count);
		for (let i: number = 0; i < goods_count; i++) {
			let xxx: GoodsItem = this._goods[i];
			packet.WriteBuffer(xxx.GetBuffer());
		}
		return packet;
	}

	public get goods(): GoodsItem[] {return this._goods; }
	public set goods(value: GoodsItem[]) { this._goods = value; }
}

File: data/golang/goods_list.go
package proto

import (
	"packet"
)

type GoodsList struct {
	Goods                    []*GoodsItem
}

func (this *GoodsList) Encode() []byte {
	pack := packet.NewWriteBuff(64)

	GoodsCount := uint16(len(this.Goods))
	pack.WriteUint16(GoodsCount)
	for i := uint16(0); i < GoodsCount; i++ {
		pack.WriteBytes(this.Goods[i].EncodeMsg())
	}

	return pack.Encode(uint16(2020))
}

func (this *GoodsList) EncodeMsg() []byte {
	pack := packet.NewWriteBuff(64)

	GoodsCount := uint16(len(this.Goods))
	pack.WriteUint16(GoodsCount)
	for i := uint16(0); i < GoodsCount; i++ {
		pack.WriteBytes(this.Goods[i].EncodeMsg())
	}

	return pack.ReadBytes()
}

func GoodsListDecode(pack *packet.Packet) *GoodsList {
	goodsList := &GoodsList{}

	GoodsCount := pack.ReadUint16()
	for ;GoodsCount > 0; GoodsCount-- {
		goodsList.Goods = append(goodsList.Goods, GoodsItemDecode(pack))
	}
	return goodsList
}
